src/coreclr/vm/jitinterface.cpp

```diff
@@ -1644,49 +1654,84 @@ static uint32_t ThreadLocalOffset(void* p)
     uint8_t* pOurTls = pTls[_tls_index];
     return (uint32_t)((uint8_t*)p - pOurTls);
 }
+#elif defined(TARGET_OSX)
+extern "C" void* GetThreadVarsAddress();
 
-void CEEInfo::getThreadLocalStaticBlocksInfo (CORINFO_THREAD_STATIC_BLOCKS_INFO* pInfo, bool isGCType)
+void* getThreadVarsSectionAddressFromDesc(uint8_t* p)
 {
-    CONTRACTL {
-        NOTHROW;
-        GC_NOTRIGGER;
-        MODE_PREEMPTIVE;
-    } CONTRACTL_END;
+    if (!(p[0] == 0x48 && p[1] == 0x8d && p[2] == 0x3d))
+    {
+        // The optimization is disabled if coreclr is not compiled in .so format.
+        return 0;
+    }
 
-    JIT_TO_EE_TRANSITION_LEAF();
+    // At this point, `p` contains the instruction pointer and is pointing to the above opcodes.
+    // These opcodes are patched by the dynamic linker.
+    // Move beyond the opcodes that we have already checked above.
+    p += 3;
 
-    pInfo->tlsIndex.addr = (void*)static_cast<uintptr_t>(_tls_index);
-    pInfo->tlsIndex.accessType = IAT_VALUE;
+    // The descriptor address is located at *p at this point.
+    // (p + 4) below skips the descriptor address bytes embedded in the instruction and
+    // add it to the `instruction pointer` to find out the address.
+    return *(uint32_t*)p + (p + 4);
+}
 
-    pInfo->offsetOfThreadLocalStoragePointer = offsetof(_TEB, ThreadLocalStoragePointer);
-    if (isGCType)
-    {
-        pInfo->offsetOfThreadStaticBlocks = ThreadLocalOffset(&t_GCThreadStaticBlocks);
-        pInfo->offsetOfMaxThreadStaticBlocks = ThreadLocalOffset(&t_GCMaxThreadStaticBlocks);
-    }
-    else
+void* getThreadVarsSectionAddress()
+{
+#ifdef TARGET_AMD64
+    // On x64, the address is related to rip, so, disassemble the function,
+    // read the offset, and then relative to the IP, find the final address of
+    // __thread_vars section.
+    uint8_t* p = reinterpret_cast<uint8_t*>(&GetThreadVarsAddress);
+    return getThreadVarsSectionAddressFromDesc(p);
+#else
+    return GetThreadVarsAddress();
+#endif // TARGET_AMD64
+}
+
+#else
+
+// Linux
+
+#ifdef TARGET_AMD64
+
+extern "C" void* GetTlsIndexObjectDescOffset();
+
+void* getThreadStaticDescriptor(uint8_t* p)
+{
+    if (!(p[0] == 0x66 && p[1] == 0x48 && p[2] == 0x8d && p[3] == 0x3d))
     {
-        pInfo->offsetOfThreadStaticBlocks = ThreadLocalOffset(&t_NonGCThreadStaticBlocks);
-        pInfo->offsetOfMaxThreadStaticBlocks = ThreadLocalOffset(&t_NonGCMaxThreadStaticBlocks);
+        // The optimization is disabled if coreclr is not compiled in .so format.
```

Comment:
```suggestion
        _ASSERTE(false && "Unexpected code sequence");
```
This should never fail. Should we assert it?

Reply:
It fails on single file scenario, but we always build in Release. I am assuming you are asking to keep the `ASSERTE` for catching things where we see this in `libcoreclr.so` case under checked. Will add.

Reply:
This is OSX path.

It fails for single file on Linux. A quick test did not fail on macOS for me (I have not tested the whole thing though.) Does it fail for you on macOS?

(The name of libcoreclr on macOS is libcoreclr.dylib.)

Reply:
> It fails for single file on Linux

Right, may be I was not clear, but it fails on linux and was assuming it might not have same instruction sequence for macos. The pipeline I was hitting the problem on Linux was `linux-x64 Release PortableSourceBuild`, but don't see equivalent pipeline for macos.

However, I verified on a hello world app and don't think the instruction stream is different for single file. 

<img width="692" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/9f7278f0-eb30-4dc0-a46a-c2dd93146248">

I will covert the if-check to an assert and remove equivalent code from `helperexpansion.cpp`.